---
title: "R Notebook"
output: html_notebook
---

```{r import data}
library(here)
library(tidyverse)
library(data.table)
library(speedyseq)
library(readxl)
source(here("R/utils.R"))

dpath <- here("data/processed")

# ------- bins (optional) --------
bins <- read_bins(here(dpath, "/contig_anno/clusters.tsv"), 
                  here(dpath, "/contig_anno/vambbins_RF_predictions.txt"))

# -------  contig annotations --------
checkv <- read_CheckV(here(dpath, "contig_anno/quality_summary.tsv"))
raw_taxa <- read_taxonomy(here(dpath, "taxonomy/taxonomy.tsv"))
catbat <- read_CATBAT(here(dpath, "/contig_anno/out.CAT.contig2classification_official.txt"))
vcontact <- read_vContact2(here(dpath, "/contig_anno/genome_by_genome_overview.csv"), "_NODE_")
ctganno <- list("checkv"=checkv,
                "catbat"=catbat,
                "vcontact"=vcontact,
                "taxonomy"=raw_taxa,
                "bins"=bins)

ctganno_merged <- ctganno$checkv %>% 
  left_join(ctganno$taxonomy, by = "Contig") %>% 
  left_join(ctganno$catbat, by = "contig_id") %>% 
  left_join(ctganno$vcontact, by = "contig_id") %>% 
  left_join(ctganno$bins, by = "Contig") %>% 
  mutate(rowid = Contig) %>% 
  column_to_rownames("rowid")

# ------- sample metadata ---------
sample_metadata <- read_excel(here(dpath, "metadata.xlsx"))

# ------- abundance table ----------
raw_abundance <- fread(here(dpath, "abundance/abundance_contigs_count.tsv"))
```


```{r Create phyloseq object}
phy_count <- raw_abundance %>% 
  column_to_rownames("Contig") %>% 
  as.matrix() %>% 
  otu_table(taxa_are_rows = T)

phy_meta <- sample_metadata %>% 
  column_to_rownames("sampleid") %>% 
  sample_data()

phy_taxa <- ctganno_merged %>% 
  as.matrix() %>%
  tax_table()

pseq <- phyloseq(phy_count, phy_taxa, phy_meta)
save(ctganno, ctganno_merged, pseq, file = here("data/vpf.RData"))
```

